bothan-core/src/manager/crypto_asset_info/price/tasks.rs

```diff
@@ -0,0 +1,156 @@
+use std::collections::VecDeque;
+
+use rust_decimal::Decimal;
+use thiserror::Error;
+use tracing::{error, info, warn};
+
+use crate::manager::crypto_asset_info::price::cache::PriceCache;
+use crate::manager::crypto_asset_info::types::{PriceState, WorkerMap};
+use crate::registry::post_processor::{PostProcess, PostProcessError};
+use crate::registry::processor::{Process, ProcessError};
+use crate::registry::signal::Signal;
+use crate::registry::source::OperationRoute;
+use crate::registry::{Registry, Valid};
+use crate::worker::AssetState;
+
+#[derive(Debug, Error)]
+enum Error {
+    #[error("Signal does not exist")]
+    InvalidSignal,
+
+    #[error("Prerequisites required: {0:?}")]
+    PrerequisiteRequired(Vec<String>),
+
+    #[error("Failed to process signal: {0}")]
+    FailedToProcessSignal(#[from] ProcessError),
+
+    #[error("Failed to post process signal: {0}")]
+    FailedToProcessPostSignal(#[from] PostProcessError),
+}
+
+pub async fn get_signal_price_states<'a>(
+    ids: Vec<String>,
+    workers: &WorkerMap<'a>,
+    registry: &Registry<Valid>,
+    stale_cutoff: i64,
+) -> Vec<PriceState> {
+    let mut cache = PriceCache::new();
+
+    let mut queue = VecDeque::from(ids.clone());
+    while let Some(id) = queue.pop_front() {
```

Comment:
can we add a counter that if we go into this loop more than, say 10k times (or whatever the reasonable number is), we log critical error and break? so we know sth bad is going on.

Reply:
I don't think this is necessary, the registry given here `Registry<Valid>` is guaranteed to not contain a cycle